cpp/src/io/comp/debrotli.cu

```diff
@@ -307,51 +309,41 @@ static __device__ uint32_t getvlc(debrotli_state_s* s, const uint16_t* lut)
   return vlc;
 }
 
+static auto __device__ allocation_size(uint32_t bytes) { return (bytes + 7) & ~7; }
+
 /// Alloc bytes from the local (shared mem) heap
-static __device__ uint8_t* local_alloc(debrotli_state_s* s, uint32_t bytes)
+static __device__ void* local_alloc(debrotli_state_s* s, uint32_t bytes)
 {
-  int heap_used = s->heap_used;
-  int len       = (bytes + 7) >> 3;
-  if (heap_used + len <= s->heap_limit) {
-    uint8_t* ptr = reinterpret_cast<uint8_t*>(&s->heap[heap_used]);
-    s->heap_used = (uint16_t)(heap_used + len);
-    return ptr;
-  } else {
-    return nullptr;
-  }
+  auto const prev_heap_used = s->heap_used;
+  auto const len            = allocation_size(bytes);
+  if (prev_heap_used + len > s->heap_limit) { return nullptr; }
+  s->heap_used = prev_heap_used + len;
+  return &s->heap[prev_heap_used];
 }
 
 /// Shrink the size of the local heap, returns ptr to end (used for stack-like intermediate
 /// allocations at the end of the heap)
-static __device__ uint8_t* local_heap_shrink(debrotli_state_s* s, uint32_t bytes)
+static __device__ void* local_heap_shrink(debrotli_state_s* s, uint32_t bytes)
 {
-  int heap_used  = s->heap_used;
-  int heap_limit = s->heap_limit;
-  int len        = (bytes + 7) >> 3;
-  if (heap_limit - len >= heap_used) {
-    heap_limit -= len;
-    s->heap_limit = (uint16_t)heap_limit;
-    return reinterpret_cast<uint8_t*>(&s->heap[heap_limit]);
-  } else {
-    return nullptr;
-  }
+  auto const len = allocation_size(bytes);
+  if (s->heap_used + len > s->heap_limit) { return nullptr; }
+  s->heap_limit = s->heap_limit - len;
```

Comment:
Do we need to make sure we don't do an unsigned rollover into a large positive value here? `len` will be uint32_t while `s->heap_limit` will be uint16_t

Reply:
I think it should be fine, since `len` is never larger than `s->heap_limit` here; we return in the previous line if `s->heap_used + len > s->heap_limit`, which is a stricter condition.